### ISSUE TITLE
推荐相关问题

### ISSUE THREAD

**Issue by 98wh** [THUDM/HGB#1]
你好，想问一下Recom/KGAT/中KGCN以及KGNNLS的实现是否与原paper不同，感觉这里的实现是类似KGAT先构建CKG, 然后学习CKG下的entity embedding? 没有找到针对KGCN这种只针对KG建模的相关代码。

**Comment by 1049451037** [comment]
您好！这里的实现原理上与KGCN和KGNNLS的paper是一样的，但是原paper是用类似链接预测的方式去评价推荐，为了适应推荐这里的训练和评价指标，就把相应的pipeline换成了KGAT里的pipeline，只将GNN模块替换成了KGCN和KGNNLS。

**Comment by 98wh** [comment]
非常感谢您的回复！我还是有些细节没有搞懂，KGCN原始实现中是不考虑user- item二部图的，只对KG进行GNN信息聚合，在您的实现中KGCN的GNN是在针对CKG进行的（def _create_gcn_embed(self):），同时KGCN中不同user应该对应不同的self.A_in，而在这个实现中貌似是不同的方式（针对CKG中每个head entity 与relation计算kgcn_score）？

**Comment by 1049451037** [comment]
我不太确定你说的不同user对应不同的A_in是指什么，KGCN论文里就是这样计算的attention score：

![image](https://user-images.githubusercontent.com/15194939/124867380-929f7180-dff0-11eb-9e94-308db0753eb5.png)

**Comment by 98wh** [comment]
我理解的KGCN中的实现是针对不同的用户分别学习不同relation的权重，然后针对同一个item，不同user u会通过不同的attention score得到不同的entity embedding  v^u。在您的实现中，对于KGCN方法是只对KG过GNN还是对CKG使用GNN？如果是CKG的处理方式，怎么保证学习到特定用户下的v^u。

**Comment by 1049451037** [comment]
我理解你的意思了～理论上确实像你说的这样，但是在这个实验里要去评测ndcg（需要对每个user-item对计算分数然后做ranking），如果用KGCN里说的batch的方法去计算user-specific item embedding计算复杂度太高了（或许这也是为什么KGCN论文里只测了AUC和F1），所以这里就直接用普通的GNN的方式实现了。

而且我们也试过直接把KGCN里对应的模块换成GCN效果其实也是一样的，所以user-specific item embedding也许没起到太大作用，可以参考：https://github.com/THUDM/HGB/tree/master/Recom/KGCN

在Recom里的KGCN实现是对CKG的。

**Comment by 98wh** [comment]
明白了，谢谢您的耐心回复～

**Comment by 1049451037** [comment]
不客气，感谢你对我们的工作的关注！

**Comment by 98wh** [comment]
您好，我还有一个关于实现推荐baseline实现细节方面的问题，在simple—HGN的实现中，bpr-mf的使用是在预测时直接额外加上e_u*e_r。我试了一下采用KGAT的那种形式发现效果与不使用bpr-mf一样差，想问一下您之前有测试过这种把pretrained embed作为初始化的方式么？
以下是我对__init__的修改：
            user_embed = pretrain['user_embed']
            item_embed = pretrain['item_embed']
            self.ini = nn.Parameter(torch.FloatTensor(np.concatenate([user_embed, item_embed], axis=0)))
            self.ret_num = user+item
            self.other_emb = nn.Parameter(torch.zeros((num_entity-self.ret_num, in_dim)))
            nn.init.xavier_normal_(self.other_emb, gain=1.414)
            self.embed = nn.Parameter(torch.cat([self.ini,self.other_emb],0))

**Comment by empty-id** [comment]
KGAT里的一些模块确实有一些设计会让结果下降，所以我们在论文里对比的时候加入了KGAT-，也就是KGAT去掉一些冗余模块的结果。

**Comment by 1049451037** [comment]
KGAT里的一些模块确实有一些设计会让结果下降，所以我们在论文里对比的时候加入了KGAT-，也就是KGAT去掉一些冗余模块的结果。

**Comment by 98wh** [comment]
感谢您的快速回复！我是在simple-HGN上进行修改的，直觉上感觉将bprmf的embed作为初始化或者预测时直接额外加上e_u*e_r两种方式应该对结果影响不大（没有修改模型的其他结构），但是目前我将simple-HGN的pretrain当作CKG中user与item  embedding初始化的话在moive-len数据集上recall从0.46下降到了0.36（ndcg类似），远小于KGAT以及KGAT-的效果。所以对这个比较迷惑，是我的处理方式有什么问题么？为什么换一种pretrain的使用方式效果会有这么大差别？

**Comment by 1049451037** [comment]
可能是使用方式有问题吧，因为Simple-HGN是经过L2 norm的，但是预训练的向量是在没有L2 norm的条件下训练的，直接用来做初始化感觉有点问题

**Comment by 98wh** [comment]
原来如此，确实有可能是因为l2导致的，谢谢～
